Computation Fields (fixes #530)

Raphcal opened the pull request:
Added a new type of field to the models. See http://www.sigmah.org/issues/view.php?id=530 for more information.

Commit:
Enable database installation through flyway maven plugin (refs #125)

Raphcal committed:
Loading countries from the cache when populated (related to #841).

Raphcal committed:
Caching the authentication to avoid rereading the cookies at each call (related to #841).

Raphcal committed:
Added an activation/deactivation mechanism for the Profiler (related to #841).

Raphcal committed:
Removed the "populate" method from the mapper and added map methods that takes an existing object instead for performance reasons (related to #841).

Raphcal committed:
First version of the computation field (related to #530) :
- added a code property on flexible elements
- fixed BatchCommand when offline
- parametrized LocalizedElement to avoid casting
- added unit tests

Raphcal committed:
Loading OpenStreetMap asynchronously to reduce load and avoid crash when OSM is unreachable (related to #841)

Raphcal committed:
Added probes at launch and when opening projects (related to #841).

Raphcal committed:
Handling computation fields on the project detail page (related to #530) +
Added the SQL script to create the required tables +
Started to check for conflicts when the computed value does not match its contraints

Raphcal committed:
Merged with upsteam/master

Raphcal committed:
Replaced Double.hashcode(...) by Double.valueOf(...).hashCode() to avoid using a Java 8 only method (related to #530)

Raphcal committed:
Mapped '×' to the multiply operator (related to #530)

Raphcal committed:
Work on the ComputationElement (related to #530) :
- Saving the code of the flexible elements in the local database.
- Keeping a reference to the objects containing the listeners to avoid losing them when offline.
- Validating the computed value.
- Replacing the element codes by the elements identifiers in the computation at save-time.

Raphcal committed:
Using the current status of the model when parsing the rule of computation elements in the administration (related to #530).

Raphcal committed:
Added a grid containing the other flexible element in the computation edit popup to show and edit their code (related to #530).

Raphcal committed:
Displaying the saved value for the minium and maximum constraints (related to #530).

Raphcal committed:
Using I18N in views related to Computation fields (related to #530) +
Displaying a message when a computed value has a bad value.

Raphcal committed:
Catching errors when the formula of a computation is unparsable (related to #530).

Raphcal committed:
Added the Jacoco Maven plugin (related to #842).

Raphcal committed:
Replaced '#' by '$' for identifiers (related to #530).

Raphcal committed:
Verifing computation constraints server side when updating a computed value (related to #530).

Raphcal committed:
Removed commented out code and empty if statement (related to #641).

Raphcal committed:
Checking computation conflicts at save time when offline (related to #530).

Raphcal committed:
Formatting the result of the computations (related to #530).

Raphcal committed:
Number, Date, Text and Paragraph are now presented as fields instead of sub-fields of TextAreaElement (related to #530).

Raphcal committed:
Added a warning message when trying to delete or disable a field (related to #530) +
Generating a default code when creating a new flexible element.

Raphcal committed:
Displaying the name of the computation fields in bold in the error message (related to #530).

Raphcal committed:
Rewritten the error message displayed when a computed value is out of bounds (related to #530) +
Changed "rule" to "formula" +
Added messages in the ComputationField administration popup.

Raphcal committed:
Rewritten the error message displayed when a computed value is out of bounds when offline (related to #530).

Raphcal committed:
Severity of the warning message changed to info (related to #530).

Raphcal committed:
Added verifications when parsing computations (related to #530).

Raphcal committed:
Corrected the mail containing the conflicts to correctly show the line ends (related to #530).

Raphcal committed:
Displaying an error message in the administration if the formula is incorrect (related to #530).

Raphcal committed:
Computing the value of a computation when none is saved (related to #530) +
Removing the current value when changing a formula.

Raphcal committed:
Added container and group columns in the code grid (related to #530) +
Added unit tests for the class DoubleValue.

Raphcal committed:
Fixed a bug when the only the maximum constraint was set (related to #530) +
Added unit tests for ComputationElementDTO.

Raphcal committed:
Fixed the error message listing the bad references of a computation (related to #530).

Raphcal committed:
Fixed the behavior of the save button when a conflict is found (related to #530).

Raphcal committed:
Merge remote-tracking branch 'upstream/master' into iss530

# Conflicts:
#	src/main/java/org/sigmah/server/handler/UpdateProjectHandler.java

Raphcal committed:
Added support for ComputationFields in the OrgUnit details page (related to #530)

Raphcal committed:
Added a flyway script for computation fields (related to #530)

Raphcal committed:
Replaced /* */ comments by -- comments (related to #530).

numero-six commented on src/main/java/org/sigmah/client/computation/ComputationTriggerManager.java:
```diff
@@ -0,0 +1,263 @@
+		
+		if (element instanceof ComputationElementDTO) {
+			components.put(element, (Field<String>) component);
+            elementsWithHandlers.put(element.getId(), (ComputationElementDTO) element);
```
That's quite strange, it seems like one line every two line is indented with tabs whereas the other is with spaces. IDE tuning required ;)

Raphcal replied:
Oops, it seems my formatter was lost when switching between branches. The main indentation style used on Sigmah is tab. Should I change this file indentation to all-tabs ?

numero-six commented on src/main/java/org/sigmah/client/util/profiler/Profiler.java:
```diff
@@ -203,110 +139,64 @@ public void onSuccess(Request result) {
-	 * @param duration Duration to format.
-	 * @return The duration formatted.
+	 * @param value Value to format.
+	 * @return The formatted value.
```
`@return The formatted value` is pretty obvious, can you elaborate? Actually, just add sample output should be sufficient (id `formatDuration(123456789) will return ...`)

numero-six commented on src/main/java/org/sigmah/server/computation/ServerValueResolver.java:
```diff
@@ -0,0 +1,55 @@
+
+				values.put(id, ComputedValues.from(result));
+			} else {
+				// TODO: Should also handle DefaultFlexibleElement.
```
Can you replace the TODO with a throw new UnsupportedOperationException("not yet implemented") if it doesn't break anything, it will make things more visible if we fall in this case.

Raphcal committed:
Changed indentation to all-tabs (#641).

ghost commented:
OK @Raphcal, I've quickly reviewed the PR and except few minor details, code can be integrated quickly.

I'm pretty happy with the size and number of commits, the commit messages are explicit. Keep that way of doing things!

Nevertheless, you pull-request is a bit too big. Github keeps telling me "Sorry, we could not display the changes to this file because there were too many other changes to display." :disappointed: 

Maybe you could have done a separate pull request (from another branch of your repo?) for everything related to profiling... Well, one PR per mantis issue is a good basis but I know it's not always that simple.

I also think that the PR is too big due to the issues about line returns (and github considering you changed the whole life whereas you only changed some lines). Long story short, tuning your IDE and your git config should fix that.

I'm more concerned by the state of your public repo. You seem to have done something wrong, all the commits on your `iss530` branch are duplicated on your `master` branch (rebased something?).

ghost commented:
Also, can you remove all calls to `System.out.println` from the tests?

You could use the logging API but if you just want to follow what test is running, you shouldn't bother with adding the same piece of code at the beginning of every test method. There are simpler way (like setting JUnit log level to `DEBUG`).

Raphcal commented:
The `System.out.println(...)` in test sources are generated by my IDE. I will remove them if you think that they are useless.

Concerning my public repository, I did do a rebase. This was in order to remove some privates commits (server addresses, jenkins configuration and other files). I know I should have created a branch specifically for them but I didn't. This was the only way I found to work around this.

Raphcal committed:
Uniformized the indentation (#641) +
Throwing an UnsupportedException in the unfinished method 'generateMarkdownReport'.

Raphcal committed:
Throwing an UnsupportedException for DefaultFlexibleElement (#641).

Raphcal committed:
Removed the calls to System.out.println that displayed the test name (#641).

Raphcal committed:
Always initializing the code grid when editing/creating a ComputationElement (related to #530) +
Uniformized indentation (tabs).

Raphcal committed:
Merge remote-tracking branch 'upstream/master' into iss530

ghost commented:
Merged, thanks!
